File: xLiAd.DiagnosticLogCenter.Agent/TracingDiagnosticMethodCollection.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Reflection;
using System.Text;
using xLiAd.DiagnosticLogCenter.Agent.DiagnosticProcessors;

namespace xLiAd.DiagnosticLogCenter.Agent
{
    internal class TracingDiagnosticMethodCollection : IEnumerable<TracingDiagnosticMethod>
    {
        private readonly List<TracingDiagnosticMethod> _methods;

        public TracingDiagnosticMethodCollection(ITracingDiagnosticProcessor tracingDiagnosticProcessor)
        {
            _methods = new List<TracingDiagnosticMethod>();
            foreach (var method in tracingDiagnosticProcessor.GetType().GetMethods())
            {
                var diagnosticName = method.GetCustomAttribute<DiagnosticName>();
                if (diagnosticName == null)
                    continue;
                _methods.Add(new TracingDiagnosticMethod(tracingDiagnosticProcessor, method, diagnosticName.Name));
            }
        }

        public IEnumerator<TracingDiagnosticMethod> GetEnumerator()
        {
            return _methods.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return _methods.GetEnumerator();
        }
    }
}

File: DiagnosticLogCenter-Java/src/main/java/com/xliad/diagnosticlogcenter/agent/TracingDiagnosticMethodCollection.java
package com.xliad.diagnosticlogcenter.agent;

import com.xliad.diagnosticlogcenter.agent.diagnosticprocessors.DiagnosticName;
import com.xliad.diagnosticlogcenter.agent.diagnosticprocessors.ITracingDiagnosticProcessor;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class TracingDiagnosticMethodCollection implements Iterable<TracingDiagnosticMethod> {
    private final List<TracingDiagnosticMethod> methods = new ArrayList<>();

    public TracingDiagnosticMethodCollection(ITracingDiagnosticProcessor processor) {
        for (Method method : processor.getClass().getDeclaredMethods()) {
            DiagnosticName diagnosticName = method.getAnnotation(DiagnosticName.class);
            if (diagnosticName != null) {
                methods.add(new TracingDiagnosticMethod(processor, method, diagnosticName.value()));
            }
        }
    }

    @Override
    public Iterator<TracingDiagnosticMethod> iterator() {
        return methods.iterator();
    }
}